### PR TITLE
ocis: init at 5.0.9

### PR DESCRIPTION
**Pull request by xinyangli**
## Description of changes

Build ownCloud Infinite Scale from source. It can be built successfully, but I do not have enough time to fully test the binary at the moment.

The [EULA](https://github.com/owncloud/ocis?tab=readme-ov-file#end-user-license-agreement) applies to release tarball. Can we safely mark this package as Apache 2.0 since we are building from source?

Related: https://github.com/NixOS/nixpkgs/issues/230190

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by poperigby** [comment]
I'm getting this error on startup:

```json
{"level":"fatal","service":"idp","error":"open identifier/index.html: file does not exist","time":"2024-09-16T19:53:11-07:00","message":"Could not open index template"
```

This is my configuration:

```nix
{ config, inputs, system, ... }:

let
    cfg = config.services.ocis;
in
{
    services = {
        ocis = {
            enable = true;
            package = inputs.ocis.legacyPackages.${system}.ocis;
            environment = {
                NOTIFICATIONS_SMTP_HOST = "smtp.sendgrid.net";
                NOTIFICATIONS_SMTP_PORT = "587";
                NOTIFICATIONS_SMTP_SENDER = "ocis@haddock.cc";
                NOTIFICATIONS_SMTP_USERNAME = "apikey";
            };
            environmentFile = config.sops.templates.ocis-secrets.path;
        };
        caddy.virtualHosts."files.haddock.cc".extraConfig = ''
            import auth
            reverse_proxy :${toString cfg.port}
        '';
    };

    sops = {
        templates.ocis-secrets = {
            content = # bash
            ''
                # IDs & passwords
                GATEWAY_STORAGE_USERS_MOUNT_ID=${config.sops.placeholder."haddock/ocis/gateway_storage_users_mount_id"}
                GRAPH_APPLICATION_ID=${config.sops.placeholder."haddock/ocis/graph_application_id"}
                IDM_IDPSVC_PASSWORD=${config.sops.placeholder."haddock/ocis/idm_idpsvc_password"}
                IDM_REVASVC_PASSWORD=${config.sops.placeholder."haddock/ocis/idm_revasvc_password"}
                IDM_SVC_PASSWORD=${config.sops.placeholder."haddock/ocis/idm_svc_password"}
                OCIS_JWT_SECRET=${config.sops.placeholder."haddock/ocis/jwt_secret"}
                OCIS_LDAP_BIND_PASSWORD=${config.sops.placeholder."haddock/ocis/ldap_bind_password"}
                OCIS_MACHINE_AUTH_API_KEY=${config.sops.placeholder."haddock/ocis/machine_auth_api_key"}
                OCIS_SERVICE_ACCOUNT_ID=${config.sops.placeholder."haddock/ocis/service_account_id"}
                OCIS_SERVICE_ACCOUNT_SECRET=${config.sops.placeholder."haddock/ocis/service_account_secret"}
                OCIS_SYSTEM_USER_API_KEY=${config.sops.placeholder."haddock/ocis/system_user_api_key"}
                OCIS_SYSTEM_USER_ID=${config.sops.placeholder."haddock/ocis/system_user_id"}
                OCIS_TRANSFER_SECRET=${config.sops.placeholder."haddock/ocis/transfer_secret"}
                STORAGE_USERS_MOUNT_ID=${config.sops.placeholder."haddock/ocis/storage_users_mount_id"}
                # OIDC
                # WEB_OIDC_CLIENT_ID=${config.sops.placeholder.ocis-oidc-client-id_ocis}
                # Mail
                NOTIFICATIONS_SMTP_PASSWORD=${config.sops.placeholder.sendgrid-api-key_ocis}
            '';
            owner = cfg.user;
        };
        secrets = {
            "haddock/ocis/gateway_storage_users_mount_id".owner = cfg.user;
            "haddock/ocis/graph_application_id".owner = cfg.user;
            "haddock/ocis/jwt_secret".owner = cfg.user;
            "haddock/ocis/ldap_bind_password".owner = cfg.user;
            "haddock/ocis/machine_auth_api_key".owner = cfg.user;
            "haddock/ocis/service_account_id".owner = cfg.user;
            "haddock/ocis/service_account_secret".owner = cfg.user;
            "haddock/ocis/storage_users_mount_id".owner = cfg.user;
            "haddock/ocis/system_user_api_key".owner = cfg.user;
            "haddock/ocis/system_user_id".owner = cfg.user;
            "haddock/ocis/transfer_secret".owner = cfg.user;
            "haddock/ocis/idm_idpsvc_password".owner = cfg.user;
            "haddock/ocis/idm_revasvc_password".owner = cfg.user;
            "haddock/ocis/idm_svc_password".owner = cfg.user;
            ocis-oidc-client-id_ocis = {
                key = "haddock/ocis/oidc_client_id";
                owner = cfg.user;
            };
            ocis-oidc-client-id_authelia = {
                key = "haddock/ocis/oidc_client_id";
                owner = "authelia-haddock";
            };
            ocis-oidc-client-secret_authelia = {
                key = "haddock/ocis/oidc_client_secret/authelia";
                owner = "authelia-haddock";
            };
            sendgrid-api-key_ocis = {
                key = "common/sendgrid/api_key";
                owner = cfg.user;
            };
        };
    };
}
```

Did you forget to copy this file over?

**Comment by xinyangli** [comment]
Hi @poperigby,

I've fixed the issue by running pnpm build in the buildPhase. Could you please test it out one more time?

**Comment by poperigby** [comment]
I'm getting a hash mismatch:

```
ocis-pnpm-Progress: resolved 2049, reused 1316, downloaded 0, added 0
error: hash mismatch in fixed-output derivation '/nix/store/dc359zdj1bvhmgxyqq7s4i8vzf572fml-ocis-pnpm-deps.drv':
        likely URL: (unknown)
         specified: sha256-ojrgoyl+xBsgEsx3d3gv5Wf6ziabHoFkEpAoDZcYavo=
            got:    sha256-Aj7dSSbXvzbj/BQVEmU9VGSmCQPZNZPssdrDDvvUYF8=
```

**Comment by xinyangli** [comment]
Sorry, I'm not familiar with `pnpm.configHook`. It can built on my side with nixpkgs-review:
<details>
<summary>
</summary>

**Comment by xinyangli** [comment]
That's strange. I can't reproduce even if I remove `ocis-pnpm-deps` and rebuild. Here is the `ocis-pnpm-deps` I got:
<details>
<summary>
nix derivation show /nix/store/c81c64ihp4w7ln32ajh8mhvqvx3k25pv-ocis-pnpm-deps
</summary>

```json
{
  "/nix/store/dc359zdj1bvhmgxyqq7s4i8vzf572fml-ocis-pnpm-deps.drv": {
    "args": [
      "-e",
      "/nix/store/v6x3cs394jgqfbi0a42pam708flxaphh-default-builder.sh"
    ],
    "builder": "/nix/store/vpvy79k1qq02p1vyqjk6nb89gwhxqvyb-bash-5.2p32/bin/bash",
    "env": {
      "__structuredAttrs": "",
      "buildInputs": "",
      "builder": "/nix/store/vpvy79k1qq02p1vyqjk6nb89gwhxqvyb-bash-5.2p32/bin/bash",
      "cmakeFlags": "",
      "configureFlags": "",
      "depsBuildBuild": "",
      "depsBuildBuildPropagated": "",
      "depsBuildTarget": "",
      "depsBuildTargetPropagated": "",
      "depsHostHost": "",
      "depsHostHostPropagated": "",
      "depsTargetTarget": "",
      "depsTargetTargetPropagated": "",
      "doCheck": "",
      "doInstallCheck": "",
      "dontBuild": "1",
      "dontConfigure": "1",
      "fixupPhase": "runHook preFixup\n\n# Remove timestamp and sort the json files\nrm -rf $out/v3/tmp\nfor f in $(find $out -name \"*.json\"); do\n  jq --sort-keys \"d
el(.. | .checkedAt?)\" $f | sponge $f\ndone\n\nrunHook postFixup\n",
      "installPhase": "runHook preInstall\n\nlockfileVersion=\"$(yq -r .lockfileVersion pnpm-lock.yaml)\"\nif [[ ${lockfileVersion:0:1} -gt 9 ]]; then\n  echo \"ERROR: l
ockfileVersion $lockfileVersion in pnpm-lock.yaml is too new for the provided pnpm version 9!\"\n  exit 1\nfi\n\nexport HOME=$(mktemp -d)\npnpm config set store-dir $out
\n# Some packages produce platform dependent outputs. We do not want to cache those in the global store\npnpm config set side-effects-cache false\n# As we pin pnpm versi
ons, we don't really care about updates\npnpm config set update-notifier false\n# Run any additional pnpm configuration commands that users provide.\n\n# pnpm is going t
o warn us about using --force\n# --force allows us to fetch all dependencies including ones that aren't meant for our host platform\npnpm install \\\n    --force \\\n
 --ignore-scripts \\\n     \\\n    --frozen-lockfile\n\nrunHook postInstall\n",
      "mesonFlags": "",
      "name": "ocis-pnpm-deps",
      "nativeBuildInputs": "/nix/store/q35mwzks79767rz6agab7919c6w22s4l-nss-cacert-3.101.1 /nix/store/rwypdgzzxk9740bd01irnq708p62axah-jq-1.7.1-dev /nix/store/sww15lfpa8
m7m05rcx6fig7visb94yyk-moreutils-0.69 /nix/store/1y9cr6yng82lkfk6yq7w60x45d726ffi-pnpm-9.7.0 /nix/store/ahch83vgag79gsiyvmqh6m1w6wwv9b9s-python3.12-yq-3.4.3",
      "out": "/nix/store/c81c64ihp4w7ln32ajh8mhvqvx3k25pv-ocis-pnpm-deps",
      "outputHash": "sha256-ojrgoyl+xBsgEsx3d3gv5Wf6ziabHoFkEpAoDZcYavo=",
      "outputHashMode": "recursive",
      "outputs": "out",
      "patches": "",
      "propagatedBuildInputs": "",
      "propagatedNativeBuildInputs": "",
      "sourceRoot": "source/services/idp",
      "src": "/nix/store/ccsmdfqdv3qfp7blm16k48j0h9jmpdw9-source",
      "stdenv": "/nix/store/ipjwd2g1kbfra83kd0dglx26xv7kn50n-stdenv-linux",
      "strictDeps": "",
      "system": "x86_64-linux",
      "version": "v5.0.7"
    },
    "inputDrvs": {
      "/nix/store/1yfcpc8qbg790k9jmcr7267ixijl11vq-pnpm-9.7.0.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/6sn8q8ssw50vvski2mifkyl8fqjjhbzs-stdenv-linux.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/7sayjrlawsxwan7pgd6ghxd4ibg3kwzv-python3.12-yq-3.4.3.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/bfl71xj1zs632dydnaf2cagsw4y5i2i2-moreutils-0.69.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/lvnmaggm7nb8nndjps5lkmmcig0spdrr-bash-5.2p32.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/mfgxm2lx7akg6jilz52y2bzbzmb65dwi-source.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/ndrz7pvwjr31nirb2v16h1g5z8ngxvi5-nss-cacert-3.101.1.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/pnmywf05lvzrallwhyim293gv74pcykj-jq-1.7.1.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "dev"
        ]
      }
    },
    "inputSrcs": [
      "/nix/store/v6x3cs394jgqfbi0a42pam708flxaphh-default-builder.sh"
    ],
    "name": "ocis-pnpm-deps",
    "outputs": {
      "out": {
        "hash": "a23ae0a3297ec41b2012cc7777782fe567face269b1e81641290280d97186afa",
        "hashAlgo": "sha256",
        "method": "nar",
        "path": "/nix/store/c81c64ihp4w7ln32ajh8mhvqvx3k25pv-ocis-pnpm-deps"
      }
    },
    "system": "x86_64-linux"
  }
}
```

</details>

I guess it's a pnpm issue? I will try on another machine.

**Comment by xinyangli** [comment]
@ofborg build ocis

**Comment by xinyangli** [comment]
Tested on my machine with a minimal config: 

```nix
{ config, pkgs, ... }:
{
  services.ocis = {
    enable = true;
    package = pkgs.ocis;
    stateDir = "/var/lib/ocis";
    url = "https://drive.xinyang.life:8443";
    address = "127.0.0.1";
    port = 9200;
    configDir = "/var/lib/ocis/config";
    environment = {
      OCIS_INSECURE = "false";
      PROXY_TLS = "false";
      OCIS_LOG_LEVEL = "debug";
      OCIS_LOG_PRETTY = "true";
      WEB_HTTP_ADDR = "127.0.0.1:12345";
    };
  };
}
```

**Comment by poperigby** [comment]
Super weird. The first time I tried building after pulling from your branch, `pnpm` had a segmentation fault, the second time it worked. No hash mismatch.

**Comment by xinyangli** [comment]
@poperigby That's super weird. It's hard to diagnose if it's not reproducible.

I did a quick dive into pnpm.fetchDeps, though. It turns out that it uses [proxy environment variables](https://github.com/NixOS/nixpkgs/blob/e88358265511d3a73b30dc6e4c15579844b35280/pkgs/development/tools/pnpm/fetch-deps/default.nix#L49C9-L49C16) and [several native dependencies](https://github.com/NixOS/nixpkgs/blob/e88358265511d3a73b30dc6e4c15579844b35280/pkgs/development/tools/pnpm/fetch-deps/default.nix#L41-L47) when calculating the hash. Are you using proxies or have an overlay on these packages? That's my best guess.

**Comment by poperigby** [comment]
No, no overlays or proxies as far as I can tell.

File: pkgs/by-name/oc/ocis/package.nix [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  nativeBuildInputs = [
+    gnumake
+    nodejs
+    pnpm.configHook
+  ];
+
+  pnpmDeps = pnpm.fetchDeps {
+    inherit pname version src;
+    sourceRoot = "${src.name}/services/idp";
+    hash = "sha256-ojrgoyl+xBsgEsx3d3gv5Wf6ziabHoFkEpAoDZcYavo=";
```
Comment (Scrumplex) [inline pkgs/by-name/oc/ocis/package.nix:53]: Please rebase and regenerate hashes.

See https://github.com/NixOS/nixpkgs/pull/349093#issuecomment-2421873637

Comment (xinyangli) [reply pkgs/by-name/oc/ocis/package.nix]: @Scrumplex Done!

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/oc/ocis/package.nix [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      cp logo.svg favicon.ico $out/share/
+    '';
+    dontConfigure = true;
+    dontBuild = true;
+    dontFixup = true;
+  };
+in
+buildGoModule rec {
+  pname = "ocis";
+  version = "v5.0.7";
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/package.nix:33]:
```suggestion
  version = "5.0.7";
```

File: pkgs/by-name/oc/ocis/package.nix [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  vendorHash = null;
+
+  src = fetchFromGitHub {
+    owner = "owncloud";
+    repo = "ocis";
+    rev = version;
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/package.nix:40]:
```suggestion
    rev = "refs/tags/v${version}";
```

File: pkgs/by-name/oc/ocis/web.nix [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    mkdir -p $out/share
+    cp -r dist/* $out/share/
+  '';
+
+  pnpmDeps = pnpm.fetchDeps {
+    inherit pname version src;
+    hash = "sha256-3Erva6srdkX1YQ727trx34Ufx524nz19MUyaDQToz6M=";
+  };
+
+  meta = with lib; {
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/web.nix:38]: Needs to be completed (you can reuse parts of ocis.meta).

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/oc/ocis/web.nix (hunk 4 above)
Comment (xinyangli) [reply pkgs/by-name/oc/ocis/web.nix:38]: > Needs to be completed (you can reuse parts of ocis.meta).


Oh, sorry about that, I forgot to stage the change 😢

**Comment by xinyangli** [comment]
- Add updateScript
- Bump to 5.0.9 from 5.0.7

File: pkgs/by-name/oc/ocis/package.nix [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  '';
+
+  passthru = {
+    web = callPackage ./web.nix { };
+    updateScript = ./update.sh;
+  };
+
+  meta = {
+    homepage = "https://github.com/owncloud/web";
+    description = "Next generation frontend for ownCloud Infinite Scale ";
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/package.nix:83]:
```suggestion
    description = "Next generation frontend for ownCloud Infinite Scale";
```

File: pkgs/by-name/oc/ocis/web.nix [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    rev = "refs/tags/v${version}";
+    hash = "sha256-hupdtK/V74+X7/eXoDmUjFvSKuhnoOtNQz7o6TLJXG4=";
+  };
+
+  nativeBuildInputs = [
+    nodejs
+    pnpm.configHook
+  ];
+
+  buildPhase = ''
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/web.nix:24]: Should include runHook preBuild / postBuild.

Same below.

File: pkgs/by-name/oc/ocis/web.nix [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    inherit pname version src;
+    hash = "sha256-3Erva6srdkX1YQ727trx34Ufx524nz19MUyaDQToz6M=";
+  };
+
+  meta = {
+    homepage = "https://github.com/owncloud/ocis";
+    description = "ownCloud Infinite Scale Stack";
+    mainProgram = "ocis";
+    maintainers = with lib.maintainers; [ xinyangli ];
+    license = [ lib.licenses.agpl3Only ];
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/web.nix:43]:
```suggestion
    license = lib.licenses.agpl3Only;
```

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/oc/ocis/web.nix [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+  pnpmDeps = pnpm.fetchDeps {
+    inherit pname version src;
+    hash = "sha256-3Erva6srdkX1YQ727trx34Ufx524nz19MUyaDQToz6M=";
+  };
+
+  meta = {
+    homepage = "https://github.com/owncloud/ocis";
+    description = "ownCloud Infinite Scale Stack";
+    mainProgram = "ocis";
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/web.nix:45]:
```suggestion
```

File: pkgs/by-name/oc/ocis/package.nix [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin/
+    cp ocis/bin/ocis $out/bin/
+    runHook postInstall
+  '';
+
+  passthru = {
+    web = callPackage ./web.nix { };
```
Comment (FliegendeWurst) [inline pkgs/by-name/oc/ocis/package.nix:77]:
```suggestion
    web = ocis.web;
```

Comment (xinyangli) [reply pkgs/by-name/oc/ocis/package.nix:77]: @FliegendeWurst Sorry but I'm not familiar with this usage. I encountered infinite recursion with this suggested change applied. Should I make other changes in order to make this work?

Comment (FliegendeWurst) [reply pkgs/by-name/oc/ocis/package.nix:77]: Hmm.. I suppose you could just omit the passthru.web. Unless you really need it somewhere.

Comment (xinyangli) [reply pkgs/by-name/oc/ocis/package.nix:77]: Well, it's directly used by the code [here](https://github.com/NixOS/nixpkgs/blob/cc960b6eeeaf40964ec547ff849293a74cd4f367/pkgs/by-name/oc/ocis/package.nix#L59). Plus, it's a separate frontend that could theoretically be hosted independently, so I don't think keeping `passthru.web` does any harm..?

Comment (FliegendeWurst) [reply pkgs/by-name/oc/ocis/package.nix:77]: In that case it is okay.

**Review by FliegendeWurst — APPROVED** [review]

File: pkgs/by-name/oc/ocis/package.nix (hunk 9 above)
Comment (xinyangli) [reply pkgs/by-name/oc/ocis/package.nix:77]: Pushed one more time for the mainProgram change. Thanks for review!

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**